### PR TITLE
vulnerabilities: expose withdrawn state on vulnerabilities

### PR DESCRIPTION
**Pull request by woodruffw**
See https://github.com/pypa/pip-audit/issues/388. This exposes the underlying "withdrawn" field provided by OSV, allowing tools like `pip-audit` to filter on it.

Signed-off-by: William Woodruff <william@trailofbits.com>

### CODE REVIEW

File: warehouse/integrations/vulnerabilities/utils.py [hunk 1]
```diff
@@ -74,9 +74,9 @@ def _analyze_vulnerability(request, vulnerability_report, origin, metrics):
     try:
         vulnerability_record = _get_vuln_record(request, report, origin)
 
-        if not report.versions:
-            # No versions indicates the vulnerability is no longer considered
-            # valid, so delete it.
+        if not report.versions or report.withdrawn is not None:
+            # No versions or a withdrawn date indicates the vulnerability is no
+            # longer considered valid, so delete it.
             _delete_vuln_record(request, vulnerability_record)
```
[the remark this answers is not in the document]
Comment (woodruffw) [reply warehouse/integrations/vulnerabilities/utils.py:80]: Sounds good; I'll go with that approach instead.

[the remark this answers is not in the document]
Comment (woodruffw) [reply warehouse/integrations/vulnerabilities/utils.py]: [de99e94](https://github.com/pypi/warehouse/pull/12443/commits/de99e9416ab8b24365d8b1562d1de273ea39c043)

File: warehouse/legacy/api/json.py [hunk 2]
```diff
@@ -129,6 +129,9 @@ def _json_data(request, project, release, *, all_releases):
             "details": vulnerability_record.details,
             "summary": vulnerability_record.summary,
             "fixed_in": vulnerability_record.fixed_in,
+            "withdrawn": vulnerability_record.withdrawn.strftime("%Y-%m-%dT%H:%M:%SZ")
+            if vulnerability_record.withdrawn
+            else None,
```
Comment (merwok) [inline warehouse/legacy/api/json.py:134]:
```suggestion
            "withdrawn": (vulnerability_record.withdrawn.strftime("%Y-%m-%dT%H:%M:%SZ")
                          if vulnerability_record.withdrawn
                          else None),
```

**Review by di — COMMENTED** [review]
LGTM, can you add an example with this field to https://warehouse.pypa.io/api-reference/json.html#known-vulnerabilities as well?

**Comment by woodruffw** [comment]
Done! Let me know what you think.

**Comment by di** [comment]
@woodruffw You've got linting failures here.

**Comment by woodruffw** [comment]
Whoops. Should be fixed now.

**Review by di — APPROVED** [review]

**Pull request merged**